xla/stream_executor/integrations/tf_allocator_adapter.h

```diff
@@ -75,56 +77,74 @@ class TfAllocatorAdapter : public DeviceMemoryAllocator {
+        per_device_allocators.resize(device_ordinal + 1);
       }
-      CHECK(!per_device_allocators_[device_ordinal]);
-      per_device_allocators_[device_ordinal] =
-          std::make_unique<TfAllocatorAdapter>(p.first.get(), p.second);
-      tf_allocators_.push_back(std::move(p.first));
-    }
-  }
-
-  MultiDeviceAdapter(const Platform *platform,
-                     std::vector<AllocatorWithLogicalIdAndStream> tf_allocators)
-      : DeviceMemoryAllocator(platform) {
-    tf_allocators_.reserve(tf_allocators.size());
-    for (AllocatorWithLogicalIdAndStream &t : tf_allocators) {
-      const int device_ordinal = std::get<1>(t);
-      Stream *stream = std::get<2>(t);
-      if (per_device_allocators_.size() <= device_ordinal) {
-        per_device_allocators_.resize(device_ordinal + 1);
-      }
-      CHECK(!per_device_allocators_[device_ordinal]);
-      per_device_allocators_[device_ordinal] =
-          std::make_unique<TfAllocatorAdapter>(std::get<0>(t).get(), stream);
-      tf_allocators_.push_back(std::move(std::get<0>(t)));
+      CHECK(!per_device_allocators[device_ordinal]);
+      per_device_allocators[device_ordinal] =
+          std::make_unique<TfAllocatorAdapter>(info.allocator.get(),
+                                               info.stream);
+      tf_allocators_.push_back(std::move(info.allocator));
     }
   }
 
   tsl::StatusOr<OwningDeviceMemory> Allocate(int device_ordinal, uint64_t size,
                                              bool retry_on_failure,
                                              int64_t memory_space) override {
-    CHECK_LT(device_ordinal, per_device_allocators_.size());
-    return per_device_allocators_[device_ordinal]->Allocate(
-        device_ordinal, size, retry_on_failure, memory_space);
+    // memory_space is used here to select allocator. This isn't a need to pass
+    // it any lower to TfAllocatorAdapter.
+    auto it = memory_space_to_per_device_allocators_.find(memory_space);
+    CHECK(it != memory_space_to_per_device_allocators_.end());
+    CHECK_LT(device_ordinal, it->second.size());
+    TF_ASSIGN_OR_RETURN(
+        auto result, it->second[device_ordinal]->Allocate(
+                         device_ordinal, size, retry_on_failure, memory_space));
+
+    absl::MutexLock lock(&mu_);
+    buffer_memory_spaces_[result->opaque()] = memory_space;
+    return result;
   }
 
   tsl::Status Deallocate(int device_ordinal, DeviceMemoryBase mem) override {
-    CHECK_LT(device_ordinal, per_device_allocators_.size());
-    return per_device_allocators_[device_ordinal]->Deallocate(device_ordinal,
-                                                              mem);
+    // Memory space is not passed to deallocate, look up in
+    // buffer_memory_spaces_.
+    int64_t memory_space;
+    {
+      absl::MutexLock lock(&mu_);
+      auto it = buffer_memory_spaces_.find(mem.opaque());
+      CHECK(it != buffer_memory_spaces_.end());
```

Comment:
One of our internal benchmarks are failing on this check.
The problem is that this method can be called for nullptr with buffer size 0.
I'll try to create small reproducible HLO, but previous code was tolerant to deallocating nullptr, maybe it's ok to add explicit check here.

Reply:
Thanks, done.

Reply:
Should we also remove this entry from buffer_memory_spaces_ for the case if the same address later will be allocated again?